### PR TITLE
Add project-request, self-provisioner, managed-resource capabilities

### PR DESCRIPTION
**Pull request by mjudeikis**
```release-note
Add the ability to exclude certain objects from azure controller reconciliation loop.
Add project-request template into openshift namespace and enable customer-admin to change it
Add ability to modify self-provisioner clusterrolebinding for cluster-admin. 
Add ability to modify shared resources in openshift namespace
```

* Deploy `project-request` into the `openshift` namespace (default location to keep aligned with the product).
* Add label `"azure.penshift.io/reconcile-protect"` to enable "deploy-once" behaviour so the customer can edit the template and we don't touch it consequence runs.
* Ability to modify cluster role allows changing role-binding from `system:authenticated` to anything else hence enabled them to deny all users to create projects. 

Some designs considerations for this change:
* I didn't wanted to add any external API changes, so we could decouple these features from MSFT needing to do any work
* most of openshift default cluster roles are protected with `openshift.io/reconcile-protect` - had to remove some protection to enable editing.
* Instead of trying to invent something new tried to keep existing OCP behaviour as close to the on-prem product as possible. 
* We dont have a clean solution to "create objects and let them to be deleted after that but still have them in sync DB". This is why I introduce new annotation, where it allows object deployment but ignores all future updates. If at any point in time we will need to update these objects, we will have to write custom logic to keep customer logic and update ours. 

Desing: https://docs.google.com/document/d/1nFq61Hdw8qXwXanLWJwEpXWFLtUvsuw4reZI4r0JghM/edit 

TODO:
* documentation on MSFT side will be needed in additional PR when we go live.

### CODE REVIEW

File: pkg/sync/v6/sync.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ const (
 	ownedBySyncPodLabelKey            = "azure.openshift.io/owned-by-sync-pod"
 	syncPodWaitForReadinessLabelKey   = "azure.openshift.io/sync-pod-wait-for-readiness"
 	syncPodReadinessPathAnnotationKey = "azure.openshift.io/sync-pod-readiness-path"
+	deployOnceSyncPodLabelKey         = "azure.openshift.io/deploy-once"
```
Comment (jim-minter) [inline pkg/sync/v6/sync.go]: it's an AnnotationKey

File: pkg/sync/v6/client.go [hunk 2]
```diff
@@ -233,6 +233,12 @@ func needsUpdate(log *logrus.Entry, existing, o *unstructured.Unstructured) bool
 		return false
 	}
 
+	// check if object is marked deploy-only and deploy once
+	if o.GetAnnotations()[deployOnceSyncPodLabelKey] != "" {
```
Comment (jim-minter) [inline pkg/sync/v6/client.go]: can you have a think now about how this mechanism could be improved to in the future support objects that we might allow the end user to delete - e.g. templates in the openshift namespace?  If we can do that now, it might be useful.

File: pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml [hunk 3]
```diff
@@ -4,6 +4,7 @@ metadata:
   annotations:
     authorization.openshift.io/system-only: "true"
     openshift.io/description: A user that can request projects.
+    openshift.io/reconcile-protect: "true"
```
Comment (jim-minter) [inline pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml]: why this change, out of interest?

File: pkg/sync/v6/sync.go (hunk 1 above)
Comment (jim-minter) [reply pkg/sync/v6/sync.go]: the deploy-once name is somewhat edge-triggered - is it possible to find an level-triggered name?

File: pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml (hunk 3 above)
Comment (mjudeikis) [reply pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml]: Because internal OCP reconcile loop keeps defaulting them back :/ we need to remove it from the reconcide loop 
https://access.redhat.com/solutions/3304341

File: pkg/sync/v6/client.go (hunk 2 above)
Comment (mjudeikis) [reply pkg/sync/v6/client.go]: Azure Arch document already has topic for this.

File: pkg/sync/v6/sync.go (hunk 1 above)
Comment (mjudeikis) [reply pkg/sync/v6/sync.go]: I might need an explanation what do you have in mind?

**Comment by mjudeikis** [comment]
We still have a question about to enable the customer to manage their `openshift` namespace.

A suggestion would be:
* enable the customer to patch `openshift` namespace so they could add same annotation `azure.openshift.io/reconcile-protect: "true"`.
* if label is added we stop updating anything in that namespace.

**Comment by mjudeikis** [comment]
/hold 
need some minor change

File: pkg/sync/v6/sync.go [hunk 4]
```diff
@@ -329,6 +338,13 @@ func (s *sync) writeDB() error {
 	if err := s.applyResources(cfgFilter, keys); err != nil {
 		return err
 	}
+	// if SharedResources are managed - update/create
+	if s.config.managedSharedResource {
+		s.log.Debug("applying storageClass resources")
```
Comment (ehashman) [inline pkg/sync/v6/sync.go]: This was copied from below I think, has nothing to do with storageClass though.

```suggestion
		s.log.Debug("applying managedShared resource")
```

File: pkg/sync/v6/client.go [hunk 5]
```diff
@@ -254,3 +259,24 @@ func printDiff(log *logrus.Entry, existing, o *unstructured.Unstructured) bool {
 	}
 	return diffShown
 }
+
+// reconcileProtected check if object is marked with azure reconcile protection annotation
+// this will remove certain objects from being updated post cluster creation
+// access to modify these objects is limited by RBAC
+func (s *sync) reconcileProtected(existing, o *unstructured.Unstructured) bool {
+	if o.GetAnnotations()[reconcileProtectKeyAnnotationKey] != "" {
```
Comment (ehashman) [inline pkg/sync/v6/client.go]: If the value is not null but e.g. "false", this will still kick in.. do we want to explicitly check that it's true?

File: pkg/sync/v6/sync.go [hunk 6]
```diff
@@ -481,6 +497,19 @@ type sync struct {
...
 	grs        []*discovery.APIGroupResources
+
+	config config
+}
+
+// config defines sync pod behaviour based on conditional logic in the cluster.
+// if set to false - components becomes unmanaged.
+// we need to make sure this for the nested dependencies (example: resource in the namespace)
+// we set these flags right time in the sync process
+type config struct {
```
Comment (ehashman) [inline pkg/sync/v6/sync.go]: `config` seems like much too generic of a name to me, especially when we're including it in a struct that has a number of other different configs already. Can we give this a more targeted name?

File: test/e2e/specs/customer_admin.go [hunk 7]
```diff
@@ -81,6 +81,50 @@ var _ = Describe("Openshift on Azure customer-admin e2e tests [CustomerAdmin][Fa
...
+
+		crb.Subjects = subject
+
+		_, err = sanity.Checker.Client.CustomerAdmin.RbacV1.ClusterRoleBindings().Update(crb)
+		Expect(err).ToNot(HaveOccurred())
+
+		// HACK: openshift reconciler should not touch this role because of:
+		// openshift.io/reconcile-protect: "true"
+		// we need to confirm this
+		time.Sleep(5 * time.Second)
```
Comment (ehashman) [inline test/e2e/specs/customer_admin.go]: Why is this sleep necessary? The comment doesn't make this any clearer to me.

File: test/e2e/specs/customer_admin.go [hunk 8]
```diff
@@ -81,6 +81,50 @@ var _ = Describe("Openshift on Azure customer-admin e2e tests [CustomerAdmin][Fa
...
+	})
+
+	It("should be able to edit self-provisioners clusterrolebinding", func() {
+		crbOriginal, err := sanity.Checker.Client.CustomerAdmin.RbacV1.ClusterRoleBindings().Get("self-provisioners", metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		crb := crbOriginal.DeepCopy()
+
+		// create ClusterRoleBinding which would allow us to test edit
+		// and would not break any other tests
+		subject := []rbacv1.Subject{
```
Comment (ehashman) [inline test/e2e/specs/customer_admin.go]:
```suggestion
		subjects := []rbacv1.Subject{
```

File: test/e2e/specs/customer_admin.go [hunk 9]
```diff
@@ -81,6 +81,50 @@ var _ = Describe("Openshift on Azure customer-admin e2e tests [CustomerAdmin][Fa
...
+		Expect(err).ToNot(HaveOccurred())
+
+		// HACK: openshift reconciler should not touch this role because of:
+		// openshift.io/reconcile-protect: "true"
+		// we need to confirm this
+		time.Sleep(5 * time.Second)
+
+		crbUpdated, err := sanity.Checker.Client.CustomerAdmin.RbacV1.ClusterRoleBindings().Get("self-provisioners", metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(len(crbUpdated.Subjects)).To(Equal(len(crb.Subjects)))
```
Comment (ehashman) [inline test/e2e/specs/customer_admin.go]: Do we want to check them one by one for equality? We didn't actually demonstrate these lengths weren't equal before, so it's possible this test is passing by coincidence.

File: pkg/sync/v6/sync.go (hunk 6 above)
Comment (mjudeikis) [reply pkg/sync/v6/sync.go]: reconcileConfig

File: test/e2e/specs/customer_admin.go (hunk 7 above)
Comment (mjudeikis) [reply test/e2e/specs/customer_admin.go]: This cole by default is managed by Openshift core rbac reconciler loop. we exclude it. This gives time for the controller to react and check if "it needs to change it". Our expectation is that we gonna manage it from now on. So we just checking if it does not touch it.

File: test/e2e/specs/customer_admin.go (hunk 9 above)
Comment (mjudeikis) [reply test/e2e/specs/customer_admin.go]: This makes sure this flow works:
	// Flow:
	// 1. Read rolebinding with default system:authenticated:oauth value (single value by default)
	// 2. Update it with 2 values (group + system:authenticated:oauth) and our reconcile annotation
	// 3. Commit to API
	// 4. Confirm what we committed is still in api/etcd

File: pkg/sync/v6/client.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/sync/v6/client.go]: done

**Comment by mjudeikis** [comment]
Added some testing notes into https://docs.google.com/document/d/1nFq61Hdw8qXwXanLWJwEpXWFLtUvsuw4reZI4r0JghM/edit#

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
```
Comment (jim-minter) [inline pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: does this actually do anything?

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  verbs:
+  - update
+  - patch
+  - delete
+  - create
+- apiGroups:
+  - ""
+  - image.openshift.io
+  resources:
+  - imagestreams/layers
```
Comment (jim-minter) [inline pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: nit: can be rolled into stanza above?

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  - create
+- apiGroups:
+  - ""
+  - project.openshift.io
+  resources:
+  - projects
+  - namespaces
+  verbs:
+  - get
+  - update
```
Comment (jim-minter) [inline pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: security: are you sure that this doesn't allow a customer admin to patch an arbitrary `openshift-*` project?  Probably an e2e to prove it would be a good plan.

File: pkg/sync/v6/client.go [hunk 13]
```diff
@@ -254,3 +259,24 @@ func printDiff(log *logrus.Entry, existing, o *unstructured.Unstructured) bool {
...
+	if strings.ToLower(o.GetAnnotations()[reconcileProtectKeyAnnotationKey]) == "true" {
+		s.log.Infof("Sync reconcile protection detected for [%s/%s], ignore update", o.GetKind(), o.GetName())
+		return true
+	}
+	if strings.ToLower(existing.GetAnnotations()[reconcileProtectKeyAnnotationKey]) == "true" {
+		s.log.Infof("User reconcile protection detected for [%s/%s], ignore update", o.GetKind(), o.GetName())
+		// check sync pod managed/unmanaged dependencies
+		if (existing.GroupVersionKind().GroupKind() == schema.GroupKind{Kind: "Namespace"} &&
+			existing.GetName() == "openshift") {
+			s.reconcileConfig.managedSharedResource = false
```
Comment (jim-minter) [inline pkg/sync/v6/client.go]: a) the lifecycle of s.reconcileConfig.managedSharedResource is going to confuse someone in the future.  Can you just explicitly get the namespace and inspect it in writeDB() and then decide whether to call applyResources(sharedResourceFilter) as a result?  That saves the pseudo-global spaghetti.
b) I think I'd rather see 2 annotations - one which prevents reconciliation of a given object, and a second one which is applied to the openshift project which prevents reconciliation of all objects within.  I don't think it makes sense to condense these into one, but am open to being persuaded.

File: pkg/sync/v6/sync.go [hunk 14]
```diff
@@ -43,6 +43,7 @@ const (
 	ownedBySyncPodLabelKey            = "azure.openshift.io/owned-by-sync-pod"
 	syncPodWaitForReadinessLabelKey   = "azure.openshift.io/sync-pod-wait-for-readiness"
 	syncPodReadinessPathAnnotationKey = "azure.openshift.io/sync-pod-readiness-path"
+	reconcileProtectKeyAnnotationKey  = "azure.openshift.io/reconcile-protect"
```
Comment (jim-minter) [inline pkg/sync/v6/sync.go]: prefer name reconcileProtectAnnotationKey

File: pkg/sync/v6/sync.go [hunk 15]
```diff
@@ -275,20 +276,25 @@ var (
 		return o.GroupVersionKind().GroupKind() == schema.GroupKind{Kind: "Secret"} ||
 			o.GroupVersionKind().GroupKind() == schema.GroupKind{Kind: "ConfigMap"}
 	}
+	sharedResourceFilter = func(o unstructured.Unstructured) bool {
+		// if Template OR ImageStream in openshift namespace
+		return o.GetNamespace() == "openshift" &&
+			(o.GroupVersionKind().GroupKind() == schema.GroupKind{Group: "image.openshift.io", Kind: "ImageStream"} ||
```
Comment (jim-minter) [inline pkg/sync/v6/sync.go]: are you sure we don't need `schema.GroupKind{Kind: "ImageStream"} and schema.GroupKind{Kind: "Template"}` as well?

File: pkg/sync/v6/client.go [hunk 16]
```diff
@@ -254,3 +259,24 @@ func printDiff(log *logrus.Entry, existing, o *unstructured.Unstructured) bool {
 	}
 	return diffShown
 }
+
+// reconcileProtected check if object is marked with azure reconcile protection annotation
+// this will remove certain objects from being updated post cluster creation
+// access to modify these objects is limited by RBAC
+func (s *sync) reconcileProtected(existing, o *unstructured.Unstructured) bool {
+	if strings.ToLower(o.GetAnnotations()[reconcileProtectKeyAnnotationKey]) == "true" {
```
Comment (jim-minter) [inline pkg/sync/v6/client.go]: isn't this a logic error?  Isn't the point of reconcileProtectKeyAnnotationKey that it's in the user's control?  We wouldn't ever want to see it in our yaml files?

File: pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml (hunk 3 above)
Comment (jim-minter) [reply pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml]: still don't get it - is this happening on clusters today?  If not, what changes here to make it a problem?

File: test/e2e/specs/customer_admin.go (hunk 7 above)
Comment (jim-minter) [reply test/e2e/specs/customer_admin.go]: +1 with @ehashman - I thought this reconcile happened at apiserver start?  Do you see an issue without this sleep?

File: test/e2e/specs/customer_admin.go [hunk 17]
```diff
@@ -81,6 +81,57 @@ var _ = Describe("Openshift on Azure customer-admin e2e tests [CustomerAdmin][Fa
...
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	// Flow:
+	// 1. Read rolebinding with default system:authenticated:oauth value
+	// 2. Update it with 2 values (group + system:authenticated:oauth)
+	// 3. Commit to API
+	// 4. Confirm what we committed is still in api/etcd
+	It("should be able to edit self-provisioners clusterrolebinding", func() {
+		crbOriginal, err := sanity.Checker.Client.CustomerAdmin.RbacV1.ClusterRoleBindings().Get("self-provisioners", metav1.GetOptions{})
```
Comment (jim-minter) [inline test/e2e/specs/customer_admin.go]: I am trying to understand if there is a scenario in which customer admin can lock themselves out.  If customer admin removes everyone from self-provisioner, is it able to add anyone back in?

File: pkg/sync/v6/client.go (hunk 16 above)
Comment (mjudeikis) [reply pkg/sync/v6/client.go]: It's an open question. We dont use it now, but I see potential use-cases in the future. But we can add them as we add the use-cases :)

File: pkg/sync/v6/client.go (hunk 13 above)
Comment (mjudeikis) [reply pkg/sync/v6/client.go]: a) It makes logic little bit complex too, as now before applying `ImageStream` or `Template` I have to inspect each object to check which namespace it is in. So It can be done, but it will be case where we have to inspect all objects as as much as `sync` care its just `unstructured`. Im ok to change this. It looked cleaner this way in my head :)
b) I don't want to create 2 annotations as this one just does one thing. And in theory, it is the same thing. :man_shrugging:

File: pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml (hunk 3 above)
Comment (mjudeikis) [reply pkg/sync/v6/data/ClusterRole.rbac.authorization.k8s.io/self-provisioner.yaml]: yes, it had to be on ClusterRoleBinding only. Not here. but it was 2 weeks when I was working with it. WIll retest and remove if this is a case.

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml (hunk 10 above)
Comment (mjudeikis) [reply pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: no

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml (hunk 12 above)
Comment (mjudeikis) [reply pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: Yes,  but its role and its assigned only to Openshift namespace. I think we would be testing RBAC here. But ok if we want to be sure our config is right.

File: pkg/sync/v6/sync.go (hunk 15 above)
Comment (mjudeikis) [reply pkg/sync/v6/sync.go]: all the current object complies with the current one. But need to confirm,

File: test/e2e/specs/customer_admin.go (hunk 7 above)
Comment (mjudeikis) [reply test/e2e/specs/customer_admin.go]: I see that sometimes our process runs faster than apiserver reacts to `watch` event and patches/not patches the object. In theroy we could read the object and see - "all good, this is what we submitted in" but in next half-second reconcile runs and changes it to default back. This  is my understanding of the process how it work (controller and reconcile loop). If this is not a case - I need to update my knowledge :)

File: pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml (hunk 12 above)
Comment (mjudeikis) [reply pkg/sync/v6/data/Role.rbac.authorization.k8s.io/openshift/shared-resource-editor.yaml]: TODO: Write e2e test for config

**Comment by mjudeikis** [comment]
TODO: move to `openshift.io/reconcile-protect: "true"`

**Comment by mjudeikis** [comment]
TODO: Add sync pod to check annotation only on self-provisioner rolebinding and openshift namespace template imagestreams

File: pkg/sync/v6/sync.go [hunk 18]
```diff
@@ -302,6 +310,20 @@ var (
...
+	// Reconcile protection logic is reconcileProtected func,
+	// but we check if openshift namespace has protection annotation
+	// to check if we need to apply sharedResourceFilter or not
+	// default - resources are managed
+	managedSharedResources := true
+	namespace, err := s.kc.CoreV1().Namespaces().Get("openshift", metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if strings.ToLower(namespace.GetAnnotations()[reconcileProtectAnnotationKey]) == "true" {
```
Comment (jim-minter) [inline pkg/sync/v6/sync.go]: are you missing a `namespace != nil &&` here?

Comment (mjudeikis) [reply pkg/sync/v6/sync.go]: `openshift` namespace exists with the start of static pods. But we can add something like that just to be sure

**Comment by ehashman** [comment]
/retest

quota flake

There's a bunch of dead jobs from this PR but I dunno which is which... perhaps we need to go purge them out a little early.

**Comment by ehashman** [comment]
/retest

flake #552
